text/0000-cfg-oneof.md

```diff
@@ -0,0 +1,96 @@
+This allows authors of crates with exclusive feature sets to communicate this to consumers without requiring users to infer the feature issue from compiler errors.
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+Additions to existing [guide](https://doc.rust-lang.org/reference/conditional-compilation.html)
+
+``` 
+ConfigurationOneof
+   oneof ( ConfigurationPredicateList )
+   
+...
+
+oneof() with a comma separated list of configuration predicates. It is true if at one and only one predicate is true. In all other situations it is false.
+```
+
+Using an existing implementation from [rust-rand-facade](https://github.com/ryankurte/rust-rand-facade/blob/master/src/lib.rs) expressing all possible combinations for a set of three exclusive features (`std`, `cortex_m`, and `os_rng`).
+
+Using existing predicates:
+
+```rust
+#[cfg(all(feature = "std", feature = "cortex_m"))]
+compile_error!("Only one of 'std', 'os_rng', or 'cortex_m' features may be enabled");
+
+#[cfg(all(feature = "std", feature = "os_rng"))]
+compile_error!("Only one of 'std', 'os_rng', or 'cortex_m' features may be enabled");
+
+#[cfg(all(feature = "cortex_m", feature = "os_rng"))]
+compile_error!("Only one of 'std', 'os_rng', or 'cortex_m' features may be enabled");
+
+#[cfg(not(any(feature = "std", feature = "cortex_m", feature = "os_rng")))]
+compile_error!("One of 'os_rng', 'std', 'cortex_m' features must be enabled");
+```
+
+Using the `oneof` predicate:
+```rust
+#[cfg(not(oneof(feature = "std", feature = "cortex_m", feature = "os_rng")))]
+compile_error!("One of 'os_rng', 'std', 'cortex_m' features must be enabled");
+```
+
+The latter is significantly more simple, and does not introduce additional complexity when adding further exclusive feature(s).
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+- I do not believe this has significant interaction with other features, or requires any significant changes to the tooling or language
+- Implementation should be straightforward, requiring the addition of a config predicate to `rustc` based on existing configuration predicates
+- I do not believe there are significant corner cases to be considered
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+N/A
+
+# Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+- This provides a useful improvement to the expressiveness of `#[cfg()]` macros and allows crate authors to provide actionable errors rather than depending on compiler failures
+- An alternative to `not(oneof(...))` could be `notoneof(...)` or similar, however, this is less generally useful and less consistent with existing predicates
+- Not implementing this means crate authors must either hand-define all variants or rely on symbol errors to enforce feature exclusivity, and users must continue to infer the _cause_ of these failures rather than being directly signalled.
```

Comment:
```suggestion
- Not implementing this means crate authors must either hand-define all variants or rely on symbol errors to enforce feature exclusivity, and users must continue to infer the _cause_ of these failures rather than being directly signaled.
```

Reply:
> and users must continue to infer the _cause_ of these failures rather than being directly signalled.

I don't follow how the existence of `oneof` provides this. The crate author must still choose to write it.

Reply:
Sure, but they/we then have the tools to reasonably do so. Do you have a suggestion for how you would reword this?